### ISSUE TITLE
ClientConfiguration gets the wrong default region from profile

### ISSUE THREAD

**Issue by owbone** [aws/aws-sdk-cpp#1310]
I have some code which tries to get the default region from an AWS profile:

```cpp
auto profile_name = Aws::Auth::GetConfigProfileName();
Aws::Client::ClientConfiguration config(profile_name.c_str());
return config.region;
```

This works fine when the region was set via `aws configure`. However, I'm trying to run this on a machine which has an instance role assigned via AWS Systems Manager. Unlike `aws configure`, SSM stores the region in `.aws/credentials` rather than `.aws/config`:

```bash
root@machine:~# cat /root/.aws/credentials 
[default]
aws_access_key_id = <...>
aws_secret_access_key = <...>
aws_session_token =  <...>
region = eu-west-1
```

The AWS CLI is able to read this region correctly:

```bash
root@machine:~# aws configure get region
eu-west-1
```

However, my code above fails to get the correct region, instead returning the default `us-east-1` all the time. Trace logs are pasted below.

### What platform/OS are you using?

Linux 97b358c342b8 4.15.0-72-generic #81-Ubuntu SMP Tue Nov 26 12:20:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

### Which version of the SDK?

1.7.264

### What compiler are you using? what version?

gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0

### Can you provide a TRACE level log? (sanitize any sensitive information)

```
[TRACE] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Checking HOME for the home directory.
[DEBUG] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Environment value for variable HOME is /root
[DEBUG] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Home directory is missing the final / appending one to normalize
[DEBUG] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Final Home Directory is /root/
[INFO] 2020-01-23 11:48:13.623 Aws::Config::AWSConfigFileProfileConfigLoader [140332231743360] Initializing config loader against fileName /root/.aws/credentials and using profilePrefix = 0
[TRACE] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Checking HOME for the home directory.
[DEBUG] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Environment value for variable HOME is /root
[DEBUG] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Home directory is missing the final / appending one to normalize
[DEBUG] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Final Home Directory is /root/
[TRACE] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Checking HOME for the home directory.
[DEBUG] 2020-01-23 11:48:13.623 FileSystemUtils [140332231743360] Environment value for variable HOME is /root
[DEBUG] 2020-01-23 11:48:13.624 FileSystemUtils [140332231743360] Home directory is missing the final / appending one to normalize
[DEBUG] 2020-01-23 11:48:13.624 FileSystemUtils [140332231743360] Final Home Directory is /root/
[INFO] 2020-01-23 11:48:13.624 ProfileConfigFileAWSCredentialsProvider [140332231743360] Setting provider to read credentials from /root/.aws/credentials for credentials file and /root/.aws/config for the config file , for use with profile default
[DEBUG] 2020-01-23 11:48:13.624 ClientConfiguration [140332231743360] ClientConfiguration will use SDK Auto Resolved profile: [default] if not specified by users.
[INFO] 2020-01-23 11:48:13.624 CurlHandleContainer [140332231743360] Initializing CurlHandleContainer with size 25
[DEBUG] 2020-01-23 11:48:13.624 ClientConfiguration [140332231743360] ClientConfiguration will use SDK Auto Resolved profile: [default] if not specified by users.
[WARN] 2020-01-23 11:48:13.624 ClientConfiguration [140332231743360] User specified profile: [default] is not found, will use the SDK resolved one.
[DEBUG] 2020-01-23 11:48:13.624 Aws::Config::ConfigFileProfileFSM [140332231743360] found profile default
[DEBUG] 2020-01-23 11:48:13.624 Aws::Config::ConfigFileProfileFSM [140332231743360] found region eu-west-1
[DEBUG] 2020-01-23 11:48:13.624 Aws::Config::ConfigFileProfileFSM [140332231743360] found access key <...>
[INFO] 2020-01-23 11:48:13.624 Aws::Config::AWSProfileConfigLoader [140332231743360] Successfully reloaded configuration.
[TRACE] 2020-01-23 11:48:13.624 Aws::Config::AWSProfileConfigLoader [140332231743360] reloaded config at 2020-01-23T11:48:13Z
```

**Comment by owbone** [comment]
I can get the correct region from the credentials with:

```cpp
Aws::Config::GetCachedCredentialsProfile(profile_name).GetRegion();
```

botocore's behaviour is to read from `.aws/config` first and then overwrite that with the values from `.aws/credentials` ([link](https://github.com/boto/botocore/blob/fd4f54bf06daad44cf92e92044a39cbd23fb0d25/botocore/session.py#L342)). Interestingly the comment there says "we don't actually need the values in the creds file", but that's not true in this case.

So, the alternative method to match the botocore behaviour:

```cpp
auto profile_name = Aws::Auth::GetConfigProfileName();
auto region = Aws::Config::GetCachedConfigProfile(profile_name).GetRegion();

if (region.empty())
{
    region = Aws::Config::GetCachedCredentialsProfile(profile_name).GetRegion();
}
```

**Comment by KaibaLopez** [comment]
Hi @owbone , 
yea, we are currently working on the credentials provider chain and how to handle default cases, and this should be fixed along with that on our next major version release.
But you're right, as is now, the provider chain the sdk uses behaves differently from boto and cli.
When calling, `Aws::Auth::GetConfigProfileName();` it only looks for the profile name on the environment variables, and if they are not set, it sets it as "default". 
And if you then use `Aws::Client::ClientConfiguration config(profile_name.c_str());`
it only looks for the "default" profile in the .aws/config file:
```
ClientConfiguration::ClientConfiguration(const char* profileName) : ClientConfiguration()
{
    if (profileName && Aws::Config::HasCachedConfigProfile(profileName))
    {
...

```
Just to give you a bit more options you could also try assigning a provider to the s3client for example:
```
auto creds = Aws::MakeShared<Aws::Auth::ProfileConfigFileAWSCredentialsProvider>("provider",profile_name.c_str());
Aws::S3::S3Client clS3Client(creds,clientConfig);
```
but honestly your work around seems fine for what you're trying to do, so hopefully this is not a blocking problem for you at the moment.
Let me know if you have other questions about this.

**Comment by KaibaLopez** [comment]
This issue has been resolved since version 1.8 I think, I'll mark it for closure for now, if it is not the case please open a new issue so that we can better keep track of.
